# Transactions Fraud Detection

**Authors:** [Peter Macinec](https://github.com/pmacinec), [Timotej Zatko](https://github.com/timzatko)

## Preprocessing

### Setup and reading the data

In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import sys
sys.path.append('..')

import pandas as pd
import numpy as np

from sklearn.pipeline import make_pipeline

from src.preprocessing.transformers import *
from src.preprocessing.pandas_feature_union import PandasFeatureUnion
from src.preprocessing.pandas_one_hot_encoder import PandasOneHotEncoder
from src.preprocessing.pandas_simple_imputer import PandasSimpleImputer
from src.preprocessing.pandas_missing_indicator import PandasMissingIndicator

from src.dataset import load_data, save_and_split_processed_data

/opt/conda/lib/python3.7/site-packages/sklearn/externals/joblib/__init__.py:15: FutureWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=FutureWarning)


In [3]:
# show full pandas output when using df.describe()

import pandas as pd

pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

In [4]:
df_transactions = load_data()

In [5]:
df_transactions.describe()

TransactionID  TransactionDT  TransactionAmt          addr2     dist1    dist2        C1        C2        C3        C4        C5        C6        C7        C8        C9       C10       C11       C12       C13            C14        D1        D2        D3        D4        D5       D6       D7            D8            D9       D10       D11      D12      D13      D14            D15             V1             V2             V3             V4             V5             V6             V7             V8             V9            V10            V11            V12            V13            V14            V15            V16            V17            V18            V19            V20            V21            V22            V23            V24            V25            V26            V27            V28            V29            V30            V31            V32            V33            V34            V35            V36            V37           V38            V39            V40  \
count   5.905400e+05   5.905400e+05   590540.000000  524834.000000  238269.0  37627.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.0  590540.000000  589271.0  309743.0  327662.0  421618.0  280699.0  73187.0  38917.0  74926.000000  74926.000000  514518.0  311253.0  64717.0  61952.0  62187.0  501427.000000  311253.000000  311253.000000  311253.000000  311253.000000  311253.000000  311253.000000  311253.000000  311253.000000  311253.000000  311253.000000  311253.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  514467.000000  421571.000000  421571.000000  421571.000000  421571.00000  421571.000000  421571.000000   
mean    3.282270e+06   7.372311e+06             NaN      86.800630       NaN      inf       NaN       NaN       0.0       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       8.295215       NaN       NaN       NaN       NaN       NaN      NaN      inf           NaN      0.000000       NaN       NaN      inf      inf      inf     163.744579       0.999945       1.045204       1.078075       0.846456       0.876991       1.045686       1.072870       1.027704       1.041529       0.463915       0.478987       0.559711       0.599166       0.999500       0.122342       0.123460       0.134040       0.135363       0.816371       0.847843       0.129684       0.132292       1.034791       1.058097       0.977660       0.988040       0.000776       0.000830       0.387840       0.406436       0.140761       0.142417       0.130693       0.139154       0.542594       0.579198       1.108065       1.16240       0.166076       0.177145   
std     1.704744e+05   4.617224e+06             NaN       2.690623       NaN      inf       NaN       NaN       0.0       NaN       NaN       NaN       NaN       NaN       0.0       NaN       NaN       NaN       NaN      49.544262       NaN       NaN       NaN       NaN       NaN      NaN      inf           NaN      0.000000       NaN       NaN      inf      inf      inf     202.726660       0.007390       0.240133       0.320890       0.440053       0.475902       0.239385       0.304779       0.186069       0.226339       0.521522       0.552431       0.510533       0.532185       0.022345       0.332422       0.342841       0.364456       0.371763       0.425512       0.459451       0.339060       0.359415       0.247681       0.305485       0.185245       0.209302       0.028596       0.031133       0.510652       0.554311       0.355793       0.368278       0.340900       0.357549       0.516010       0.539039       0.690571       0.85892       0.451956       0.505786   
min     2.987000e+06   8.640000e+04        

In [6]:
df_transactions.shape

(590540, 434)

There are 590540 rows in the dataset with 434 features.

### Define pipeline

In [7]:
categoric_features = df_transactions.select_dtypes(include=np.object).columns.to_list()
numeric_features = df_transactions.select_dtypes(exclude=np.object).columns.to_list()

pipeline = PandasFeatureUnion([
    ('categoric_features', make_pipeline(
        SelectFeatures(categoric_features),
        EmailProviderTransform(['P_emaildomain', 'R_emaildomain']),    
        KeepOnlyMostCommonValues(10),
        PandasOneHotEncoder()
    )),
    ('numeric_features', make_pipeline(
        SelectFeatures(numeric_features),
        FilterColumnsByCountOfMissingValues(0.5),
        PandasSimpleImputer(strategy='mean'),
        PandasMissingIndicator()
    ))
])

In [8]:
%%time

new_df_transactions = pipeline.fit_transform(df_transactions)

../src/preprocessing/transformers.py:69: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  x[column] = x[column].astype('str').apply(transform_email).astype('str')
/opt/conda/lib/python3.7/site-packages/pandas/core/indexing.py:671: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_with_indexer(indexer, value)
../src/preprocessing/transformers.py:50: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a

CPU times: user 1min 25s, sys: 1min 8s, total: 2min 34s
Wall time: 3min 33s


In [9]:
new_df_transactions.shape

(590540, 539)

After preprocessing there are 590540 rows with 539 features.

In [10]:
%%time

save_and_split_processed_data(new_df_transactions, test_size=0.2)

splitting the data...
saving...
CPU times: user 2min 58s, sys: 8.06 s, total: 3min 6s
Wall time: 3min 58s
